src/containers/SearchPage/MainPanel.js

```diff
@@ -1,31 +1,142 @@
+  // keyword search sorts the listings according to relevance.
+  // In those cases, sort parameter should be removed.
+  const sortingFiltersActive = findActiveFilter(
+    sortConfig.conflictingFilters,
+    searchParams,
+    filterConfig
+  );
+  return sortingFiltersActive ? { ...searchParams, sort: null } : searchParams;
+};
+
+/**
+ * MainPanel contains search results and filters.
+ * There are 3 presentational container-components that show filters:
+ * SearchfiltersMobile, SearchFiltersPrimary, and SearchFiltersSecondary.
+ * The last 2 are for desktop layout.
+ */
 class MainPanel extends Component {
   constructor(props) {
     super(props);
-    this.state = { isSearchFiltersPanelOpen: false };
+    this.state = { isSecondaryFiltersOpen: false, currentQueryParams: props.urlQueryParams };
+
+    this.applyFilters = this.applyFilters.bind(this);
+    this.cancelFilters = this.cancelFilters.bind(this);
+    this.resetAll = this.resetAll.bind(this);
+
+    this.initialValues = this.initialValues.bind(this);
+    this.handleChangedValue = this.handleChangedValue.bind(this);
+
+    // SortBy
+    this.handleSortBy = this.handleSortBy.bind(this);
+  }
+
+  // Apply the filters by redirecting to SearchPage with new filters.
+  applyFilters() {
+    const { history, urlQueryParams, sortConfig, filterConfig } = this.props;
+    const searchParams = { ...urlQueryParams, ...this.state.currentQueryParams };
+    const search = cleanSearchFromConflictingParams(searchParams, sortConfig, filterConfig);
+
+    history.push(createResourceLocatorString('SearchPage', routeConfiguration(), {}, search));
+  }
+
+  // Close the filters by clicking cancel, revert to the initial params
+  cancelFilters() {
+    this.setState({ currentQueryParams: {} });
+  }
+
+  // Reset all filter query parameters
+  resetAll(e) {
+    const { urlQueryParams, history, filterConfig } = this.props;
+    const filterQueryParamNames = filterConfig.map(f => f.queryParamNames);
+
+    // Reset state
+    this.setState({ currentQueryParams: {} });
+
+    // Reset routing params
+    const queryParams = omit(urlQueryParams, filterQueryParamNames);
+    history.push(createResourceLocatorString('SearchPage', routeConfiguration(), {}, queryParams));
+  }
+
+  initialValues(queryParamNames) {
+    const currentQueryParams = this.state.currentQueryParams;
+    const urlQueryParams = this.props.urlQueryParams;
+    const isArray = Array.isArray(queryParamNames);
+
+    const getInitialValue = paramName => {
+      const currentQueryParam = currentQueryParams[paramName];
+      const hasQueryParamInState = typeof currentQueryParam !== 'undefined';
+      return hasQueryParamInState ? currentQueryParam : urlQueryParams[paramName];
+    };
+
+    return isArray
+      ? queryParamNames.reduce((acc, paramName) => {
+          return { ...acc, [paramName]: getInitialValue(paramName) };
+        }, {})
+      : {};
+  }
+
+  handleChangedValue(useHistoryPush) {
```

Comment:
@kpuputti Because it needs to have useHistory flag somewhere in the scope chain (and that flag is specific to filter instance).

Maybe it should be renamed to "getHandleChangedValue"?